### PR TITLE
Support extra data and files in a more general way

### PR DESCRIPTION
**Pull request by RobertKrawitz**
This allows specifying more complex workloads without special casing the template code.

```
metadata:
  name: myworkload
extra_files:
  - file1.yaml
  - file2.yaml
template_data:
  shared:
    ...shared_data
  run_type:
     perf_ci:
       ... perf_ci data
     default:
       ... data if no other run type matches
  kind:
    vm:
      run_type:
        perf_ci:
        default:
      default:
  extra:  # Do we need sub-structuring here?
    mariadb:
    default:

### CODE REVIEW

**Comment by RobertKrawitz** [comment]
Currently a draft PR; this is for discussion rather than immediate integration.

**Comment by RobertKrawitz** [comment]
@ebattat @jeniferh please review

**Comment by RobertKrawitz** [comment]
Passed e2e and unit tests on my local cluster.  Most of the integration tests passed except for the ones that always fail on my cluster.

**Comment by ebattat** [comment]
@RobertKrawitz,
I think that we need to change the name of the file name to more meaning name:
benchmark_runner/common/template_operations/update_data_template_yaml_with_environment_variables.py

jinja_render_environmant_variable.py => what do you think on this name ?

File: benchmark_runner/common/template_operations/template_operations.py [hunk 1]
```diff
@@ -22,107 +22,92 @@ def __initialize_dependent_variables__(self):
...
-        workload_template = self.__get_yaml_template_by_workload()
-        workload_name = self.__get_workload_name()
-        workload_dir_path = os.path.join(self.__dir_path, workload_name)
-        self.__environment_variables_dict['workload_name'] = workload_name
+        def build_template_data(previous_data: dict, template_root_data: dict):
+            """
+            Build data dictionary for rendering from template data.
+            """
+            def get_sub_dict(dictionary: dict, key: str, value: str):
+                if key in dictionary:
```
[the remark this answers is not in the document]
Comment (RobertKrawitz) [reply benchmark_runner/common/template_operations/template_operations.py:48]: It is a general parameter; this could be any dictionary.  I'll doc both.

**Comment by RobertKrawitz** [comment]
> @RobertKrawitz, I think that we need to change the name of the file name to more meaning name: benchmark_runner/common/template_operations/update_data_template_yaml_with_environment_variables.py
> 
> jinja_render_environmant_variable.py => what do you think on this name ?

Perhaps simply "render_yaml_from_template.py"?

File: benchmark_runner/common/template_operations/templates/hammerdb/hammerdb_data_template.yaml [hunk 2]
```diff
@@ -1,66 +1,73 @@
...
-        cores: 4
-        vm_limits_memory: 16Gi
-        vm_requests_memory: 10Mi
+metadata:
+  name: hammerdb
+{%- if kind != 'vm' %}
+files:
+  - name: "{{ workload_extra_name }}.yaml"
+{%- endif %}
+template_data:
```
Comment (jeniferh) [inline benchmark_runner/common/template_operations/templates/hammerdb/hammerdb_data_template.yaml:7]: I am a little confused, what is provided in this yaml?

File: benchmark_runner/common/template_operations/template_operations.py [hunk 3]
```diff
@@ -22,107 +22,92 @@ def __initialize_dependent_variables__(self):
...
-                       **kind_data, **run_type_data, **kind_runtype_data}
-        workload_file_suffix = ''
+        template_render_data = {**self.__environment_variables_dict, **common_data}
+
+        workload_data = yaml.load(render_yaml_file(workload_dir_path, f'{self.__workload_name}_data_template.yaml', template_render_data), Loader=yaml.FullLoader)
+        render_data = build_template_data(template_render_data, workload_data)
 
         answer = {}
-        if workload_name == 'hammerdb':
-            database = self.__get_workload_extra()
```
[the remark this answers is not in the document]
Comment (RobertKrawitz) [reply benchmark_runner/common/template_operations/template_operations.py:105]: It isn't called "database" any more; it's simply the extra part of the workload (`workload_extra_name`).  The template code doesn't care whether it's a database or anything else.

File: benchmark_runner/common/template_operations/template_operations.py [hunk 4]
```diff
@@ -22,107 +22,92 @@ def __initialize_dependent_variables__(self):
             self.__environment_variables_dict['es_suffix'] = '-test-ci'
         else:
             self.__environment_variables_dict['es_suffix'] = ''
-        # hammerdb storage
-        if self.__environment_variables_dict.get('ocs_pvc', '') == 'True':
-            self.__storage_type = 'ocs_pvc'
-        else:
-            self.__storage_type = 'ephemeral'
-
```
[the remark this answers is not in the document]
Comment (RobertKrawitz) [reply benchmark_runner/common/template_operations/template_operations.py:30]: Yes, that's now set in the environment variables and the templates use them.

File: benchmark_runner/common/template_operations/templates/hammerdb/hammerdb_data_template.yaml (hunk 2 above)
Comment (RobertKrawitz) [reply benchmark_runner/common/template_operations/templates/hammerdb/hammerdb_data_template.yaml:7]: `workload_extra_name` is what used to be called `database`.

A workload named `hammerdb_pod_mariadb` has a `workload_name` of `hammerdb`, `workload_kind` (and in this case `workload_template_kind`) of `pod`, and `workload_extra_name` of `mariadb`.

**Review by ebattat — APPROVED** [review]
/LGTM

**Pull request merged**